File: android/app/src/main/java/com/lilru/liftr/nutrition/NutritionMetabolism.kt
package com.lilru.liftr.nutrition

import com.lilru.liftr.data.BackendContracts
import java.time.LocalDate
import java.time.Period
import java.time.ZoneId
import kotlin.math.roundToInt

object NutritionMetabolism {

    fun sexOffset(sex: String?): Double {
        return when (sex?.trim()?.lowercase().orEmpty()) {
            "male", "m" -> BackendContracts.NutritionMetabolism.MALE_OFFSET
            "female", "f" -> BackendContracts.NutritionMetabolism.FEMALE_OFFSET
            else -> BackendContracts.NutritionMetabolism.UNISEX_OFFSET
        }
    }

    fun workoutActivityMultiplier(workoutsPerWeek: Double): Double {
        val wpw = workoutsPerWeek.coerceAtLeast(0.0)
        return when {
            wpw < 1.5 -> BackendContracts.NutritionMetabolism.MULTIPLIER_LOW
            wpw < 3.5 -> BackendContracts.NutritionMetabolism.MULTIPLIER_MODERATE
            wpw < 5.5 -> BackendContracts.NutritionMetabolism.MULTIPLIER_ACTIVE
            else -> BackendContracts.NutritionMetabolism.MULTIPLIER_VERY_ACTIVE
        }
    }

    fun imputedHeightCm(sex: String?, heightCm: Double?): Double {
        if (heightCm != null && heightCm > 0.0) return heightCm
        return when (sex?.trim()?.lowercase().orEmpty()) {
            "male", "m" -> BackendContracts.NutritionMetabolism.DEFAULT_HEIGHT_MALE_CM
            else -> BackendContracts.NutritionMetabolism.DEFAULT_HEIGHT_FEMALE_CM
        }
    }

    fun imputedWeightKg(sex: String?, weightKg: Double?): Double {
        if (weightKg != null && weightKg > 0.0) return weightKg
        return when (sex?.trim()?.lowercase().orEmpty()) {
            "male", "m" -> BackendContracts.NutritionMetabolism.DEFAULT_WEIGHT_MALE_KG
            else -> BackendContracts.NutritionMetabolism.DEFAULT_WEIGHT_FEMALE_KG
        }
    }

    fun demographicFallbackKcal(sex: String?): Int {
        return when (sex?.trim()?.lowercase().orEmpty()) {
            "female", "f" -> BackendContracts.NutritionMetabolism.FALLBACK_KCAL_FEMALE
            "male", "m" -> BackendContracts.NutritionMetabolism.FALLBACK_KCAL_MALE
            else -> BackendContracts.NutritionMetabolism.FALLBACK_KCAL_NEUTRAL
        }
    }

    fun computeBmrKcal(
        sex: String?,
        dateOfBirth: LocalDate?,
        heightCm: Double?,
        weightKg: Double?,
        today: LocalDate = LocalDate.now(ZoneId.systemDefault())
    ): Int {
        val height = imputedHeightCm(sex, heightCm)
        val weight = imputedWeightKg(sex, weightKg)
        val ageYears = if (dateOfBirth != null) {
            Period.between(dateOfBirth, today).years
        } else {
            BackendContracts.NutritionMetabolism.IMPUTED_AGE_YEARS
        }
        val bmr = BackendContracts.NutritionMetabolism.WEIGHT_FACTOR * weight +
            BackendContracts.NutritionMetabolism.HEIGHT_FACTOR * height -
            BackendContracts.NutritionMetabolism.AGE_FACTOR * ageYears +
            sexOffset(sex)
        return bmr.roundToInt().coerceIn(
            BackendContracts.NutritionMetabolism.MIN_KCAL,
            BackendContracts.NutritionMetabolism.MAX_KCAL
        )
    }

    fun computeMetabolicTargetKcal(
        sex: String?,
        dateOfBirth: LocalDate?,
        heightCm: Double?,
        weightKg: Double?,
        workoutsPerWeek: Double = 0.0,
        today: LocalDate = LocalDate.now(ZoneId.systemDefault())
    ): Int {
        val bmr = computeBmrKcal(sex, dateOfBirth, heightCm, weightKg, today)
        val tdee = bmr * workoutActivityMultiplier(workoutsPerWeek)
        return tdee.roundToInt().coerceIn(
            BackendContracts.NutritionMetabolism.MIN_KCAL,
            BackendContracts.NutritionMetabolism.MAX_KCAL
        )
    }

    fun resolveDisplayKcal(
        sex: String?,
        dateOfBirth: LocalDate?,
        heightCm: Double?,
        weightKg: Double?,
        storedTarget: Int?,
        isManual: Boolean,
        workoutsPerWeek: Double = 0.0,
        today: LocalDate = LocalDate.now(ZoneId.systemDefault())
    ): Int {
        if (isManual) {
            val stored = storedTarget ?: demographicFallbackKcal(sex)
            return stored.coerceIn(
                BackendContracts.NutritionMetabolism.MIN_KCAL,
                BackendContracts.NutritionMetabolism.MAX_KCAL
            )
        }
        return computeMetabolicTargetKcal(
            sex,
            dateOfBirth,
            heightCm,
            weightKg,
            workoutsPerWeek,
            today
        )
    }
}

File: Liftr/NutritionMetabolism.swift
import Foundation

enum NutritionMetabolism {
    static let fallbackKcalFemale = 1500
    static let fallbackKcalMale = 1900
    static let fallbackKcalNeutral = 1700
    static let minKcal = 800
    static let maxKcal = 6000
    static let imputedAgeYears = 30

    static func sexOffset(sex: String?) -> Double {
        let normalized = (sex ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        switch normalized {
        case "male", "m": return 5
        case "female", "f": return -161
        default: return -80
        }
    }

    static func workoutActivityMultiplier(workoutsPerWeek: Double) -> Double {
        let wpw = max(0, workoutsPerWeek)
        if wpw < 1.5 { return 1.2 }
        if wpw < 3.5 { return 1.375 }
        if wpw < 5.5 { return 1.55 }
        return 1.725
    }

    static func imputedHeightCm(sex: String?, heightCm: Double?) -> Double {
        if let heightCm, heightCm > 0 { return heightCm }
        let normalized = (sex ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return normalized == "male" || normalized == "m" ? 175 : 162
    }

    static func imputedWeightKg(sex: String?, weightKg: Double?) -> Double {
        if let weightKg, weightKg > 0 { return weightKg }
        let normalized = (sex ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return normalized == "male" || normalized == "m" ? 75 : 60
    }

    static func demographicFallbackKcal(sex: String?) -> Int {
        let normalized = (sex ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        switch normalized {
        case "female", "f": return fallbackKcalFemale
        case "male", "m": return fallbackKcalMale
        default: return fallbackKcalNeutral
        }
    }

    static func computeBmrKcal(
        sex: String?,
        birthDate: Date?,
        heightCm: Double?,
        weightKg: Double?,
        referenceDate: Date = Date()
    ) -> Int {
        let height = imputedHeightCm(sex: sex, heightCm: heightCm)
        let weight = imputedWeightKg(sex: sex, weightKg: weightKg)
        let ageYears: Int
        if let birthDate {
            ageYears = Calendar.current.dateComponents([.year], from: birthDate, to: referenceDate).year ?? imputedAgeYears
        } else {
            ageYears = imputedAgeYears
        }
        let bmr = (10 * weight) + (6.25 * height) - (5 * Double(ageYears)) + sexOffset(sex: sex)
        return Int(bmr.rounded()).clamped(to: minKcal...maxKcal)
    }

    static func computeMetabolicTargetKcal(
        sex: String?,
        birthDate: Date?,
        heightCm: Double?,
        weightKg: Double?,
        workoutsPerWeek: Double = 0,
        referenceDate: Date = Date()
    ) -> Int {
        let bmr = computeBmrKcal(
            sex: sex,
            birthDate: birthDate,
            heightCm: heightCm,
            weightKg: weightKg,
            referenceDate: referenceDate
        )
        let mult = workoutActivityMultiplier(workoutsPerWeek: workoutsPerWeek)
        return Int((Double(bmr) * mult).rounded()).clamped(to: minKcal...maxKcal)
    }

    static func resolveDisplayKcal(
        sex: String?,
        birthDate: Date?,
        heightCm: Double?,
        weightKg: Double?,
        storedTarget: Int?,
        isManual: Bool,
        workoutsPerWeek: Double = 0,
        referenceDate: Date = Date()
    ) -> Int {
        if isManual {
            let stored = storedTarget ?? demographicFallbackKcal(sex: sex)
            return stored.clamped(to: minKcal...maxKcal)
        }
        return computeMetabolicTargetKcal(
            sex: sex,
            birthDate: birthDate,
            heightCm: heightCm,
            weightKg: weightKg,
            workoutsPerWeek: workoutsPerWeek,
            referenceDate: referenceDate
        )
    }
}

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
